### PR TITLE
[16.0][MIG] database_cleanup: Migration to 16.0

### PR DESCRIPTION
**Pull request by JoelZilli**

### CODE REVIEW

**Review by gurneyalex — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by santostelmo — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Can you please rebase and check the tests, that seems to be failing?

**Review by StephaneMangin — APPROVED** [review]

**Comment by StephaneMangin** [comment]
cc @pedrobaeza Feel free to approve if it's ok you.

**Comment by pedrobaeza** [comment]
There's still problems in tests, and in fact, this PR doesn't contain right now any CI check. Don't know why. Please rebase again.

**Comment by StephaneMangin** [comment]
@JoelZilli Please integrate this commit to fix the CI issue.

https://github.com/yankinmax/server-tools/pull/1

EDIT: a specific PR for this version is about to be proposed, stay tuned

**Comment by gurneyalex** [comment]
@JoelZilli I can't make a PR on your branch please cherry pick this commit https://github.com/gurneyalex/server-tools/commit/0bda5bc777c40ff38f4a0be1ac8f0ba7dfad4c4a

**Comment by paulius-sladkevicius** [comment]
Got a few warning logs while testing it.

```
2023-03-28 11:45:22,085 1 WARNING odoodb odoo.api.create: The model odoo.addons.database_cleanup.models.purge_wizard is not overriding the create method in batch 
2023-03-28 11:45:22,085 1 WARNING odoodb odoo.api.create: The model odoo.addons.database_cleanup.models.purge_wizard is not overriding the create method in batch 
```

File: database_cleanup/models/purge_models.py [hunk 1]
```diff
@@ -2,34 +2,38 @@
...
 
 class IrModel(models.Model):
     _inherit = "ir.model"
 
     def _drop_table(self):
         """this function crashes for undefined models"""
-        existing_model_ids = self.filtered(lambda x: x.model in self.env)
-        return super(IrModel, existing_model_ids)._drop_table()
+        self = self.filtered(lambda x: x.model in self.env)
+        return super()._drop_table()
```
Comment (hbrunn) [inline database_cleanup/models/purge_models.py:21]: this won't work, you really need to call super only on the restricted set

File: database_cleanup/models/purge_models.py [hunk 2]
```diff
@@ -2,34 +2,38 @@
...
 
     @api.depends()
     def _inherited_models(self):
         """this function crashes for undefined models"""
-        existing_model_ids = self.filtered(lambda x: x.model in self.env)
-        super(IrModel, existing_model_ids)._inherited_models()
+        self = self.filtered(lambda x: x.model in self.env)
+        return super()._inherited_models()
```
Comment (hbrunn) [inline database_cleanup/models/purge_models.py:27]: see above

File: database_cleanup/models/purge_models.py [hunk 3]
```diff
@@ -2,34 +2,38 @@
...
 
 class IrModelFields(models.Model):
     _inherit = "ir.model.fields"
 
     def _prepare_update(self):
         """this function crashes for undefined models"""
-        existing = self.filtered(lambda x: x.model in self.env)
-        return super(IrModelFields, existing)._prepare_update()
+        self = self.filtered(lambda x: x.model in self.env)
+        return super()._prepare_update()
```
Comment (hbrunn) [inline database_cleanup/models/purge_models.py:36]: see above

**Review by hbrunn — CHANGES_REQUESTED** [review]

**Comment by hbrunn** [comment]
please cherry pick https://github.com/hbrunn/server-tools/commit/e85698cecaa8ae15b2eabc8ed437e046bac165be for fixed module purging

**Comment by JoelZilli** [comment]
On the migration of the module to 14 this "purge" (here https://github.com/OCA/server-tools/commit/b18e74e2dda7c02285e772cad1ddef080d9e111e#diff-1bd698f0fb309ad37089dfe60270acca4e447b5afa31a28aab9c358d7b9a4b72L80) call was removed and then all this lines are not making anything on the database. Instead of re-adding purge and removing modules automatically when entering the menu, we believe it's better to show uninstalled modules that are not available so that they can be removed togheter with the other modules.

**Comment by hbrunn** [comment]
that was part of the commit I asked you to cherry pick above

**Comment by JoelZilli** [comment]
Sorry, didn`t see that. I added your changes directly on my commit. Thanks!

**Comment by gurneyalex** [comment]
Hello @JoelZilli  Can you cherry pick https://github.com/OCA/server-tools/commit/e8d68eb8db12d9aeb9e3ff9e6ba9d7dcdedbf301 (adapting the branch name in test.yml) to fix the CI?

File: database_cleanup/models/create_indexes.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        for field in self.env["ir.model.fields"].search(
+            [
+                ("index", "=", True),
+            ]
+        ):
+            if field.model not in self.env.registry:
+                continue
+            model = self.env[field.model]
+            name = "{}_{}_index".format(model._table, field.name)
+            self.env.cr.execute(
```
Comment (fmdl) [inline database_cleanup/models/create_indexes.py:60]: trigram index are not found

File: database_cleanup/security/ir.model.access.csv [hunk 5]
```diff
@@ -0,0 +1,17 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_cleanup_create_indexes_line,access_cleanup_create_indexes_line,model_cleanup_create_indexes_line,base.group_user,1,1,1,1
```
Comment (fmdl) [inline database_cleanup/security/ir.model.access.csv:2]: use this group : base.group_erp_manager, no ?

File: database_cleanup/views/purge_wizard.xml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                <header>
+                    <button
+                        type="object"
+                        name="purge_all"
+                        string="Purge all"
+                        class="oe_highlight"
+                    />
+                    <button type="object" name="select_lines" string="Select lines" />
+                </header>
+                <div attrs="{'invisible': [('purge_line_ids', '!=', [])]}">
```
Comment (fmdl) [inline database_cleanup/views/purge_wizard.xml:16]: don't forget to add sheet.

File: database_cleanup/views/create_indexes.xml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    </record>
+
+    <record id="cleanup_create_indexes_line_view_tree" model="ir.ui.view">
+        <field name="model">cleanup.create_indexes.line</field>
+        <field name="inherit_id" ref="tree_purge_line" />
+        <field name="mode">primary</field>
+        <field name="arch" type="xml">
+            <button name="purge" position="attributes">
+                <attribute name="string">Create this index</attribute>
+                <attribute name="icon">gtk-add</attribute>
```
Comment (fmdl) [inline database_cleanup/views/create_indexes.xml:37]: this icon doesn't exist

File: database_cleanup/models/purge_tables.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                    self.env.cr.execute(
+                        "ALTER TABLE %s DROP CONSTRAINT %s",
+                        (
+                            IdentifierAdapter(constraint[3]),
+                            IdentifierAdapter(constraint[0]),
+                        ),
+                    )
+
+            self.logger.info("Dropping table %s", line.name)
+            self.env.cr.execute("DROP TABLE %s", (IdentifierAdapter(line.name),))
```
Comment (fmdl) [inline database_cleanup/models/purge_tables.py:75]: maybe add cascade in `DROP TABLE %s CASCADE`, tu prevent any postgres error.

**Review by fmdl — CHANGES_REQUESTED** [review]

**Comment by StephaneMangin** [comment]
@JoelZilli Could you cherry pick https://github.com/OCA/server-tools/commit/e8d68eb8db12d9aeb9e3ff9e6ba9d7dcdedbf301 please ? To fix the tests

**Comment by legalsylvain** [comment]
Hi. What is the state of this PR ? Thanks !

**Comment by pedrobaeza** [comment]
Superseded by #2684

**Pull request closed without merging**